lang/eval.go

```diff
@@ -226,22 +225,63 @@ func (s *Scope) evalContext(refs []*addrs.Reference, selfAddr addrs.Referenceabl
 				continue
 			}
 
-			// Treat "self" as an alias for the configured self address.
-			rawSubj = selfAddr
-			isSelf = true
+			// self can only be used within a resource instance
+			subj := selfAddr.(addrs.ResourceInstance)
```

Comment:
This was originally designed to be generic so that we could in principle support `self` in other contexts in future, since there are several issues open around that sort of thing. Fully generalizing this isn't necessary right now but since this limitation is not obvious in the `Scope` API, could we handle this by panicking a more specific error instead?

Relatedly, it seems like the condition right below this can no longer be `true` because if `selfAddr` were `addrs.Self` this type assertion would've panicked already. That branch is also just trying to present a more helpful panic message when the caller uses this API improperly, so perhaps we should invert this to let that one "win" for the more specific case.